\PassOptionsToPackage{utf8}{inputenc}
\documentclass{bioinfo}
\copyrightyear{2018}
\pubyear{2018}

\access{Advance Access Publication Date: Day Month Year}
\appnotes{Phylogenetics}

% ######################################################################################################################
%         Custom Packages
% ######################################################################################################################

% \usepackage[T1]{fontenc}
\usepackage[utf8]{inputenc}
\usepackage[english]{babel}

% text subscript https://tex.stackexchange.com/a/1017
% \usepackage{fixltx2e}
% \usepackage{hyperref}

% https://tex.stackexchange.com/questions/135358/changing-the-formatting-of-subcaption-for-reference
\usepackage[labelformat=simple]{subcaption}
\renewcommand\thesubfigure{(\alph{subfigure})}

% https://tex.stackexchange.com/a/6105
\usepackage[binary-units=true]{siunitx}

% https://tex.stackexchange.com/a/39981
\usepackage[nolist,nohyperlinks]{acronym}

% https://tex.stackexchange.com/a/163779
\usepackage{amsmath}
\usepackage{algorithm}
\usepackage[noend]{algpseudocode}
\makeatletter
\def\BState{\State\hskip-\ALG@thistlm}
\makeatother

% https://tex.stackexchange.com/questions/3372/how-do-i-typeset-arbitrary-fractions-like-the-standard-symbol-for-5-%C2%BD
\usepackage{xfrac}

% https://tex.stackexchange.com/a/5255
\DeclareMathOperator*{\argmax}{arg\,max}

% https://tex.stackexchange.com/a/58088
\usepackage{lmodern}

\usepackage[table,dvipsnames]{xcolor}

% Include them here again to ensure correct setup.
% See https://tex.stackexchange.com/q/1863
\usepackage[colorlinks=true, allcolors=blue]{hyperref}
\renewcommand\UrlFont{\color{black}\sffamily}

\usepackage{nameref,zref-xr}
\zxrsetup{toltxlabel}
\zexternaldocument*[supp:]{art-supp}

% https://tex.stackexchange.com/questions/5959/cool-text-highlighting-in-latex
\usepackage{soul}

% line numbers
% \usepackage[switch]{lineno}

% This is a quick fix for needless warnings on letter paper.
% Need to tinker with the geometry package in the class file to fix this!
% https://tex.stackexchange.com/a/117212
\vbadness=10000
\hbadness=10000

% Define custom styles
\newcommand\toolname{\textsc}
\newcommand\taxonname{\textit}

% ######################################################################################################################
%         Temporaries
% ######################################################################################################################

% https://tex.stackexchange.com/questions/21300/custom-counter-and-cross-referencing
\newcounter{questionno}
\newcommand\question[1]{\stepcounter{questionno} \hl{\thequestionno: #1}}
\newcommand\todo[1]{{\color{purple}{#1}}}
\newcommand\alexis[1]{{\color{orange}{#1}}}
\newcommand\nicetohave[1]{{\color{RoyalBlue}{#1}}}

\renewcommand\todo[1]{}
\renewcommand\alexis[1]{}
\renewcommand\nicetohave[1]{}
\renewcommand\question[1]{}

% Comments and envorinments from the original template.
% Might be useful if needed later.

% Citation command:
% \citealp{Boffelli03}
% \citep{Boffelli03}
% \citep{Boffelli03}
% \citet{Boffelli03}

% \begin{methods}
% \end{methods}

% please remove the " % " symbol from \centerline{\includegraphics{fig01.eps}} as it may ignore the figures.
%
% \begin{figure}[!tpb]%figure1
% % \fboxsep=0pt\colorbox{gray}{\begin{minipage}[t]{235pt} \vbox to 100pt{\vfill\hbox to
% % 235pt{\hfill\fontsize{24pt}{24pt}\selectfont FPO\hfill}\vfill}
% % \end{minipage}}
% % \centerline{\includegraphics{fig01.eps}}
% \caption{Caption, caption.}\label{fig:01}
% \end{figure}
%
% \begin{figure}[!tpb]%figure2
% %\centerline{\includegraphics{fig02.eps}}
% \caption{Caption, caption.}\label{fig:02}
% \end{figure}

% Equation~(\ref{eq:01}) Text Text Text Text Text Text  Text Text
% Text Text Text Text Text Text Text Text Text Text Text Text Text.
% Figure~2\vphantom{\ref{fig:02}}

% \begin{table}[!t]
% \processtable{This is table caption\label{Tab:01}} {\begin{tabular}{@{}llll@{}}\toprule head1 &
% head2 & head3 & head4\\\midrule
% row1 & row1 & row1 & row1\\
% row2 & row2 & row2 & row2\\
% row3 & row3 & row3 & row3\\
% row4 & row4 & row4 & row4\\\botrule
% \end{tabular}}{This is a footnote}
% \end{table}

% (Table~\ref{Tab:01}) Text Text Text Text Text Text  Text Text Text

% ######################################################################################################################
%         Title and Header
% ######################################################################################################################

\begin{document}
\firstpage{1}

% \subtitle{}
\subtitle{Subject Section}

% Short title: no more than 50 chars
\title[Automatic Reference Trees \& Multilevel Placement]{Methods for Automatic Reference Trees and Multilevel Phylogenetic Placement}
% \title[Automatic Phylogenies and Multilevel Placement]{Algorithms for Inference of Reference Phylogenies and Multilevel Phylogenetic Placement}
\author[Czech \textit{et~al.}]{Lucas Czech\,$^{\text{\sfb 1,}\ast}$, Pierre Barbera\,$^{\text{\sfb 1}}$, and Alexandros Stamatakis\,$^{\text{\sfb 1,2}}$}
\address{
$^{\text{\sf 1}}$Scientific Computing Group, Heidelberg Institute for Theoretical Studies, Heidelberg, Germany\\
$^{\text{\sf 2}}$Institute for Theoretical Informatics, Karlsruhe Institute of Technology, Karlsruhe, Germany}

\corresp{$^\ast$To whom correspondence should be addressed.}

% \history{}
% \editor{}

\history{Received on XXXXX; revised on XXXXX; accepted on XXXXX}

\editor{Associate Editor: XXXXXXX}

% ######################################################################################################################
%         Acronyms. Have to be inside the document environment...
% ######################################################################################################################

\begin{acronym}
    \acro{RT}[RT]{reference tree}
%     \acroplural{RT}[RTs]{reference trees}
    \acro{RA}[RA]{reference alignment}
%     \acroplural{RA}[RAs]{reference alignment}
    \acro{QS}[QS]{query sequence}
%     \acroplural{QS}[QSs]{query sequences}

    \acro{BT}[BT]{backbone tree}
%     \acroplural{BT}[BTs]{backbone trees}
    \acro{CT}[CT]{clade tree}
%     \acroplural{CT}[CTs]{clade trees}

    \acro{ART}[ART]{automatic reference tree}

    \acro{BV}[BV]{Bacterial Vaginosis}
    \acro{NTS}[NTS]{Neotropical Soils}
    \acro{TO}[TO]{Tara Oceans}
    \acro{HMP}[HMP]{Human Microbiome Project}
\end{acronym}

% ######################################################################################################################
%         Abstract
% ######################################################################################################################

% Max of 150 words recommended. Motivation and Results have a total of 152 currently.
\abstract{
\textbf{Motivation:}
% Needed to shorten...
% Molecular sequencing costs are decreasing exponentially, leading to unprecedented amounts of genetic data.
% In most metagenomic studies, one of the first steps is to asses the evolutionary provenance of this data.
In most metagenomic sequencing studies, the initial analysis step consists in assessing the evolutionary provenance of the sequences.
Phylogenetic (or Evolutionary) Placement methods can be employed
to determine the evolutionary position of sequences with respect to a given reference phylogeny.
These placement methods do however face certain limitations:
The manual selection of reference sequences is labor-intensive;
the computational effort to infer reference phylogenies is substantially larger than for methods that rely on sequence similarity;
the number of taxa in the reference phylogeny should be small enough to allow for visually inspecting the results.
\\
\textbf{Results:}
% USPs: no manual ref selection (fast prototyping), fast pipeline for large datasets, multilevel for better tree vis and speed.
We present algorithms to overcome the above limitations.
First, we introduce a method to automatically construct representative sequences from databases to infer reference phylogenies.
Second, we present an approach for conducting large-scale phylogenetic placements on nested phylogenies.
Third, we describe a preprocessing pipeline that allows for handling huge sequence data sets.
Our experiments on empirical data show that our methods substantially accelerate the workflow and
yield highly accurate placement results.
\\
% \textbf{Availability and Implementation:}
\textbf{Implementation:}
Freely available under GPLv3 at \url{http://github.com/lczech/gappa}.
\\
\textbf{Contact:} \href{lucas.czech@h-its.org}{lucas.czech@h-its.org}
\\
\textbf{Supplementary Information:} Supplementary data are available at \textit{Bioinformatics} online.
}

\maketitle
% \linenumbers

% ######################################################################################################################
%         Introduction
% ######################################################################################################################

\section{Introduction}
\label{sec:Introduction}


% Studies often target specific regions of the genome,
% % e.g., the 16S or 18S regions of \taxonname{Bacteria} or \taxonname{Eukaryotes},
% resulting in so-called meta-barcoding reads.
A typical task in such studies is to identify and classify the reads by
relating them to known reference sequences, either taxonomically or phylogenetically.

\nicetohave{
Conventional methods like \toolname{Blast} \citep{Altschul1990} are fast and work reasonably well
}
Conventional methods based on sequence similarity are fast and work reasonably well
if the reads are similar enough to the reference sequences,
that is, if they represent species that are closely related to known species.
\nicetohave{
However, the best \toolname{Blast} hit does often \emph{not} yield the most closely related species \citep{Koski2001}.
}
However, they might \emph{not} yield the most closely related species \citep{Koski2001}.
This is particularly true for environments where available reference databases
do not exhibit sufficient taxon coverage \citep{Mahe2017}.
As insufficient taxon coverage cannot be detected by methods that are based on sequence similarity,
they can potentially bias downstream analyses.

So-called phylogenetic (or evolutionary) placement methods \citep{Matsen2010,Berger2011,Barbera2018}
provide a more accurate means for identifying reads.
Instead of relying on sequence similarity,
they identify reads based on a phylogenetic tree of reference sequences.
Thereby, they can incorporate information about the evolutionary history of the species under study.

In short, phylogenetic placement calculates the most probable insertion branches for a \acf{QS} on a given \acf{RT}.
For metagenomic studies, the \acp{QS} are the reads from the environmental samples,
and most often barcoding regions or marker genes are used, see below.
First, the \acp{QS} are aligned against the reference alignment of the \ac{RT}.
\alexis{schaun wir mal wie lang der text wird, wenn es reicht sollte die aligner rein}
\nicetohave{
, using programs such as \toolname{PaPaRa} \citep{Berger2011a,Berger2012} or
\toolname{hmmalign}, which is a subprogram of the \toolname{HMMER} suite \citep{Eddy1998,Eddy2009}.
% PaPaRa is already cited in the supplement, when describing details about the data processing.
} % end of nice to have
For a given \ac{QS} and a given branch in the \ac{RT},
the \ac{QS} is inserted as a new tip into the branch;
the affected branch lengths are then re-optimized;
the likelihood score of the tree is evaluated;
and the \ac{QS} is removed again from the branch.
This process yields a so-called \emph{placement} of the \ac{QS} for every branch of the \ac{RT},
that is, an optimized position on the branch, along with a likelihood score for the entire \ac{RT}.
These likelihood scores are then transformed into probabilities
to quantify the uncertainty of the \ac{QS} placement into the respective branch.
\nicetohave{\citep{Strimmer2002,VonMering2007}} % one reference should be sufficient here.
This placement process is repeated independently for each \ac{QS} on the original \ac{RT}.
Phylogenetic placement thus yields a mapping of each \ac{QS} to all branches of the \ac{RT},
along with a probability for each placement of a \ac{QS} on a specific branch.
% For details, refer to \citep{Matsen2010,Berger2011}.
% That is the very condensed version of it. We don't need all the detail from the other paper here!

This mapping can be seen as an identification and classification of the \acp{QS} in terms of the \ac{RT},
similar to a taxonomic assignment.
However, phylogenetic placement also allows for more elaborate downstream analyses.
Firstly, the reference tree usually offers a higher resolution than simple abundance counts per taxon,
and the amount of mapped \acp{QS} per branch can be directly visualized on the \ac{RT} \citep{Mahe2017}.
Secondly, established methods such as Edge PCA and Squash Clustering \citep{Matsen2011a}
allow to detect subtle differences between different samples,
thus enabling comparative studies based on phylogenetic placement.
Lastly, we recently proposed novel methods for visualizing and clustering phylogenetic placement data \citep{Czech2018a},
which for example can reveal correlations of per-sample meta-data features with the sequence abundances.

Phylogenetic placement is applicable if the \acp{QS} can be aligned to the reference alignment.
Often, barcoding regions such as 16S or 18S are used,
but there also haven been studies using different maker genes \citep{Sunagawa2013a}.
Furthermore, other types of sequences such as $_{\text{mi}}$tags %\textsubscript{mi}tags
\citep{Logares2014} can be used.
Phylogenetic placement is particularly helpful for studying new, unexplored environments,
for which no closely related sequences exist in reference databases (e.g., \citealp{Mahe2017}).
However, the selection of suitable reference sequences for inferring the \ac{RT} constitutes
a challenge for studying such environments, as it is often a manual process.
Furthermore, conducting phylogenetic placements requires a higher computational effort
with respect to the placement algorithms \emph{per se}, but also the pre- and post-processing,
than, for instance, similarity based methods.
Nonetheless, existing placement algorithms are being increasingly used and cited.
Due to the continuous advances in molecular sequencing, existing placement methods
as well as respective pre- and post-processing tools have already reached their scalability limits.
% Another limitation is the selection process of reference sequences.

% Evolutionary Placement can be used for taxonomic assignment.
% In studies that specifically look for certain kinds of organisms,
% e.g, protists \citep{Mahe2017},
% it usually suffices to use a taxonomy covering the organisms of interest,
% potentially including outgroups from more distance species.
% As metagenomic analyses get cheaper,
% it is however to be expected that researchers want to target more than one group of organisms within one study.
% Thus, this limitation needs to be overcome.

% ######################################################################################################################
%         Methods
% ######################################################################################################################

\section{Methods}
\label{sec:Methods}

% ======================================================================================================================
%     Automatic Reference Trees
% ======================================================================================================================

\subsection{Automatic Reference Trees}
\label{sec:Methods:sub:AutomaticTrees}

% ======================================================================================================================
%     Multilevel Placement
% ======================================================================================================================

\subsection{Multilevel Placement}
\label{sec:Methods::sub:MultilevelPlacement}


% ======================================================================================================================
%     Phylogenetic Placement
% ======================================================================================================================

\subsection{Data Preprocessing for Phylogenetic Placement}
\label{sec:Methods:sub:DataPreprocessing}

% ######################################################################################################################
%         Results
% ######################################################################################################################

\section{Results}
\label{sec:Results}


% ######################################################################################################################
%         Discussion
% ######################################################################################################################

\section{Discussion and Conclusion}
\label{sec:DiscussionConclusion}


% ######################################################################################################################
%         Acknowledgements
% ######################################################################################################################

% \vspace*{-12pt}
\section*{Funding}
\label{sec:Funding}

This work was financially supported by the \textbf{Klaus Tschira Stiftung gGmbH} in Heidelberg, Germany.
% \vspace*{-12pt}

\section*{Acknowledgements}
\label{sec:Acknowledgements}

% \vspace*{-12pt}

% ######################################################################################################################
%         References
% ######################################################################################################################

\bibliographystyle{natbib}
%\bibliographystyle{achemnat}
%\bibliographystyle{plainnat}
%\bibliographystyle{abbrv}
%\bibliographystyle{bioinformatics}
%
%\bibliographystyle{plain}
%
\bibliography{art-bib}


% \begin{thebibliography}{}

% \bibitem[Bofelli {\it et~al}., 2000]{Boffelli03}
% Bofelli,F., Name2, Name3 (2003) Article title, {\it Journal Name}, {\bf 199}, 133-154.

% \bibitem[Bag {\it et~al}., 2001]{Bag01}
% Bag,M., Name2, Name3 (2001) Article title, {\it Journal Name}, {\bf 99}, 33-54.

% \bibitem[Yoo \textit{et~al}., 2003]{Yoo03}
% Yoo,M.S. \textit{et~al}. (2003) Oxidative stress regulated genes
% in nigral dopaminergic neurnol cell: correlation with the known
% pathology in Parkinson's disease. \textit{Brain Res. Mol. Brain
% Res.}, \textbf{110}(Suppl. 1), 76--84.

% \bibitem[Lehmann, 1986]{Leh86}
% Lehmann,E.L. (1986) Chapter title. \textit{Book Title}. Vol.~1, 2nd edn. Springer-Verlag, New York.

% \bibitem[Crenshaw and Jones, 2003]{Cre03}
% Crenshaw, B.,III, and Jones, W.B.,Jr (2003) The future of clinical
% cancer management: one tumor, one chip. \textit{Bioinformatics},
% doi:10.1093/bioinformatics/btn000.

% \bibitem[Auhtor \textit{et~al}. (2000)]{Aut00}
% Auhtor,A.B. \textit{et~al}. (2000) Chapter title. In Smith, A.C.
% (ed.), \textit{Book Title}, 2nd edn. Publisher, Location, Vol. 1, pp.
% ???--???.

% \bibitem[Bardet, 1920]{Bar20}
% Bardet, G. (1920) Sur un syndrome d'obesite infantile avec
% polydactylie et retinite pigmentaire (contribution a l'etude des
% formes cliniques de l'obesite hypophysaire). PhD Thesis, name of
% institution, Paris, France.

% \end{thebibliography}
\end{document}
