% Template for PLoS
% Version 3.4 January 2017
%
% % % % % % % % % % % % % % % % % % % % % %
%
% -- IMPORTANT NOTE
%
% This template contains comments intended
% to minimize problems and delays during our production
% process. Please follow the template instructions
% whenever possible.
%
% % % % % % % % % % % % % % % % % % % % % % %
%
% Once your paper is accepted for publication,
% PLEASE REMOVE ALL TRACKED CHANGES in this file
% and leave only the final text of your manuscript.
% PLOS recommends the use of latexdiff to track changes during review, as this will help to maintain a clean tex file.
% Visit https://www.ctan.org/pkg/latexdiff?lang=en for info or contact us at latex@plos.org.
%
%
% There are no restrictions on package use within the LaTeX files except that
% no packages listed in the template may be deleted.
%
% Please do not include colors or graphics in the text.
%
% The manuscript LaTeX source should be contained within a single file (do not use \input, \externaldocument, or similar commands).
%
% % % % % % % % % % % % % % % % % % % % % % %
%
% -- FIGURES AND TABLES
%
% Please include tables/figure captions directly after the paragraph where they are first cited in the text.
%
% DO NOT INCLUDE GRAPHICS IN YOUR MANUSCRIPT
% - Figures should be uploaded separately from your manuscript file.
% - Figures generated using LaTeX should be extracted and removed from the PDF before submission.
% - Figures containing multiple panels/subfigures must be combined into one image file before submission.
% For figure citations, please use "Fig" instead of "Figure".
% See http://journals.plos.org/plosone/s/figures for PLOS figure guidelines.
%
% Tables should be cell-based and may not contain:
% - spacing/line breaks within cells to alter layout or alignment
% - do not nest tabular environments (no tabular environments within tabular environments)
% - no graphics or colored text (cell background color/shading OK)
% See http://journals.plos.org/plosone/s/tables for table guidelines.
%
% For tables that exceed the width of the text column, use the adjustwidth environment as illustrated in the example table in text below.
%
% % % % % % % % % % % % % % % % % % % % % % % %
%
% -- EQUATIONS, MATH SYMBOLS, SUBSCRIPTS, AND SUPERSCRIPTS
%
% IMPORTANT
% Below are a few tips to help format your equations and other special characters according to our specifications. For more tips to help reduce the possibility of formatting errors during conversion, please see our LaTeX guidelines at http://journals.plos.org/plosone/s/latex
%
% For inline equations, please be sure to include all portions of an equation in the math environment.  For example, x$^2$ is incorrect; this should be formatted as $x^2$ (or $\mathrm{x}^2$ if the romanized font is desired).
%
% Do not include text that is not math in the math environment. For example, CO2 should be written as CO\textsubscript{2} instead of CO$_2$.
%
% Please add line breaks to long display equations when possible in order to fit size of the column.
%
% For inline equations, please do not include punctuation (commas, etc) within the math environment unless this is part of the equation.
%
% When adding superscript or subscripts outside of brackets/braces, please group using {}.  For example, change "[U(D,E,\gamma)]^2" to "{[U(D,E,\gamma)]}^2".
%
% Do not use \cal for caligraphic font.  Instead, use \mathcal{}
%
% % % % % % % % % % % % % % % % % % % % % % % %
%
% Please contact latex@plos.org with any questions.
%
% % % % % % % % % % % % % % % % % % % % % % % %

\documentclass[10pt,letterpaper]{article}
\usepackage[top=0.85in,left=2.75in,footskip=0.75in]{geometry}

% amsmath and amssymb packages, useful for mathematical formulas and symbols
\usepackage{amsmath,amssymb}

% Use adjustwidth environment to exceed column width (see example table in text)
\usepackage{changepage}

% Use Unicode characters when possible
\usepackage[utf8x]{inputenc}

% textcomp package and marvosym package for additional characters
\usepackage{textcomp,marvosym}

% cite package, to clean up citations in the main text. Do not remove.
\usepackage{cite}

% Use nameref to cite supporting information files (see Supporting Information section for more info)
\usepackage{nameref,hyperref}

% line numbers
\usepackage[right]{lineno}

% ligatures disabled
\usepackage{microtype}
\DisableLigatures[f]{encoding = *, family = * }

% color can be used to apply background shading to table cells only
\usepackage[table]{xcolor}

% array package and thick rules for tables
\usepackage{array}

% create "+" rule type for thick vertical lines
\newcolumntype{+}{!{\vrule width 2pt}}

% create \thickcline for thick horizontal lines of variable length
\newlength\savedwidth
\newcommand\thickcline[1]{%
  \noalign{\global\savedwidth\arrayrulewidth\global\arrayrulewidth 2pt}%
  \cline{#1}%
  \noalign{\vskip\arrayrulewidth}%
  \noalign{\global\arrayrulewidth\savedwidth}%
}

% \thickhline command for thick horizontal lines that span the table
\newcommand\thickhline{\noalign{\global\savedwidth\arrayrulewidth\global\arrayrulewidth 2pt}%
\hline
\noalign{\global\arrayrulewidth\savedwidth}}


% Remove comment for double spacing
%\usepackage{setspace}
%\doublespacing

% Text layout
\raggedright
\setlength{\parindent}{0.5cm}
\textwidth 5.25in
\textheight 8.75in

% Bold the 'Figure #' in the caption and separate it from the title/caption with a period
% Captions will be left justified
\usepackage[aboveskip=1pt,labelfont=bf,labelsep=period,justification=raggedright,singlelinecheck=off]{caption}
\renewcommand{\figurename}{Fig}

% Use the PLoS provided BiBTeX style
\bibliographystyle{plos2015}

% Remove brackets from numbering in List of References
\makeatletter
\renewcommand{\@biblabel}[1]{\quad#1.}
\makeatother

% Leave date blank
\date{}

% Header and Footer with logo
\usepackage{lastpage,fancyhdr,graphicx}
\usepackage{epstopdf}
\pagestyle{myheadings}
\pagestyle{fancy}
\fancyhf{}
\setlength{\headheight}{27.023pt}
% \lhead{\includegraphics[width=2.0in]{PLOS-submission.eps}}
\rfoot{\thepage/\pageref{LastPage}}
\renewcommand{\footrule}{\hrule height 2pt \vspace{2mm}}
\fancyheadoffset[L]{2.25in}
\fancyfootoffset[L]{2.25in}
% \lfoot{\sf PLOS}

%% Include all macros below

\newcommand{\lorem}{{\bf LOREM}}
\newcommand{\ipsum}{{\bf IPSUM}}

%% END MACROS SECTION

% ######################################################################################################################
%         Custom Packages
% ######################################################################################################################

% https://tex.stackexchange.com/questions/135358/changing-the-formatting-of-subcaption-for-reference
\usepackage[labelformat=simple]{subcaption}
\renewcommand\thesubfigure{(\alph{subfigure})}

% https://tex.stackexchange.com/a/6105
\usepackage[binary-units=true]{siunitx}

% https://tex.stackexchange.com/a/39981
\usepackage[nolist,nohyperlinks]{acronym}

% https://tex.stackexchange.com/a/163779
\usepackage{amsmath}
\usepackage{algorithm}
\usepackage[noend]{algpseudocode}
\makeatletter
\def\BState{\State\hskip-\ALG@thistlm}
\makeatother

% https://tex.stackexchange.com/questions/3372/how-do-i-typeset-arbitrary-fractions-like-the-standard-symbol-for-5-%C2%BD
\usepackage{xfrac}

\usepackage{booktabs}
% https://tex.stackexchange.com/questions/94845/problems-with-toprule-and-midrule-in-a-table

% https://tex.stackexchange.com/a/5255
\DeclareMathOperator*{\argmax}{arg\,max}

% https://tex.stackexchange.com/a/58088
\usepackage{lmodern}

% https://tex.stackexchange.com/a/195599
\usepackage[T1]{fontenc}
\usepackage{textcomp}

% https://tex.stackexchange.com/questions/192449/how-to-change-spacing-between-caption-numbering-and-text
\DeclareCaptionLabelFormat{supplement_fig}{S#1#2 Fig. }
\DeclareCaptionLabelFormat{supplement_tab}{S#1#2 Table. }
% \DeclareCaptionLabelSeparator{supplabelsep}{}

% https://tex.stackexchange.com/questions/62611/how-to-make-ref-references-include-the-word-figure
\usepackage{cleveref}

% https://tex.stackexchange.com/q/139463
\usepackage{soul}

% text subscript https://tex.stackexchange.com/a/1017
\usepackage{fixltx2e}
\usepackage{hyperref}

% http://bytesizebio.net/2013/03/11/adding-supplementary-tables-and-figures-in-latex/
\newcommand{\beginsupplement}{%
    \setcounter{table}{0}
%     \renewcommand{\thetable}{S\arabic{table}}%
    \renewcommand{\tablename}{}%
    \captionsetup[table]{labelformat=supplement_tab,labelsep=none}
    \setcounter{figure}{0}
%     \renewcommand{\thefigure}{\arabic{figure}}%
    \renewcommand{\figurename}{}
    \captionsetup[figure]{labelformat=supplement_fig,labelsep=none}
}

\newcommand\figref[1]{Fig~\ref{#1}}
\newcommand\suppfigref[1]{S\ref{#1}~Fig}

% Define custom styles
\newcommand\toolname{\textsc}
\newcommand\taxonname{\textit}

% ######################################################################################################################
%         Temporaries
% ######################################################################################################################

% https://tex.stackexchange.com/questions/21300/custom-counter-and-cross-referencing
% \newcounter{questionno}
% \newcommand\question[1]{{\stepcounter{questionno} \hl{\thequestionno: #1}}}
% \newcommand\question[1]{{\hl{#1}}}
\newcommand\todo[1]{{\color{purple}{#1}}}
\newcommand\alexis[1]{{\color{orange}{#1}}}
% \newcommand\todo[1]{}
% \newcommand\alexis[1]{}

% \usepackage{pdfcomment}
% \newcommand\redbarbegin{\begin{pdfsidelinecomment}[color=red,linewidth=2bp]{}}
% \newcommand\redbarend{\end{pdfsidelinecomment}}

% Comments and envorinments from the original template.
% Might be useful if needed later.

% PLOS does not support heading levels beyond the 3rd (no 4th level headings).

% For figure citations, please use "Fig" instead of "Figure".
% Nulla mi mi, Fig~\ref{fig1} venenatis, \nameref{S1_Video}

% Place figure captions after the first paragraph in which they are cited.
% \begin{figure}[!h]
% \caption{{\bf Bold the figure title.}
% Figure caption text here, please use this space for the figure panel descriptions instead of using subfigure commands. A: Lorem ipsum dolor sit amet. B: Consectetur adipiscing elit.}
% \label{fig1}
% \end{figure}

% Use "Eq" instead of "Equation" for equation citations.
% \begin{eqnarray}
% \label{eq:schemeP}
%     \mathrm{P_Y} = \underbrace{H(Y_n) - H(Y_n|\mathbf{V}^{Y}_{n})}_{S_Y} + \underbrace{H(Y_n|\mathbf{V}^{Y}_{n})- H(Y_n|\mathbf{V}^{X,Y}_{n})}_{T_{X\rightarrow Y}},
% \end{eqnarray}

% \begin{enumerate}
%     \item{react}
%     \item{diffuse free particles}
%     \item{increment time by dt and go to 1}
% \end{enumerate}

% \begin{itemize}
%     \item First bulleted item.
%     \item Second bulleted item.
%     \item Third bulleted item.
% \end{itemize}

% % Place tables after the first paragraph in which they are cited.
% \begin{table}[!ht]
% \begin{adjustwidth}{-2.25in}{0in} % Comment out/remove adjustwidth environment if table fits in text column.
% \centering
% \caption{
% {\bf Table caption Nulla mi mi, venenatis sed ipsum varius, volutpat euismod diam.}}
% \begin{tabular}{|l+l|l|l|l|l|l|l|}
% \hline
% \multicolumn{4}{|l|}{\bf Heading1} & \multicolumn{4}{|l|}{\bf Heading2}\\ \thickhline
% $cell1 row1$ & cell2 row 1 & cell3 row 1 & cell4 row 1 & cell5 row 1 & cell6 row 1 & cell7 row 1 & cell8 row 1\\ \hline
% $cell1 row2$ & cell2 row 2 & cell3 row 2 & cell4 row 2 & cell5 row 2 & cell6 row 2 & cell7 row 2 & cell8 row 2\\ \hline
% $cell1 row3$ & cell2 row 3 & cell3 row 3 & cell4 row 3 & cell5 row 3 & cell6 row 3 & cell7 row 3 & cell8 row 3\\ \hline
% \end{tabular}
% \begin{flushleft} Table notes Phasellus venenatis, tortor nec vestibulum mattis, massa tortor interdum felis, nec pellentesque metus tortor nec nisl. Ut ornare mauris tellus, vel dapibus arcu suscipit sed.
% \end{flushleft}
% \label{table1}
% \end{adjustwidth}
% \end{table}

% ######################################################################################################################
%         Title and Header
% ######################################################################################################################

\begin{document}
\vspace*{0.2in}

% Title must be 250 characters or less.
\begin{flushleft}
{\Large
\textbf\newline{Scalable methods for post-processing, visualizing, and analyzing phylogenetic placements}
% Please use "sentence case" for title and headings (capitalize only the first word in a title (or heading), the first word in a subtitle (or subheading), and any proper nouns).
}
\newline
% Insert author names, affiliations and corresponding author email (do not include titles, positions, or degrees).
\\
Lucas Czech\textsuperscript{1,*},
Alexandros Stamatakis\textsuperscript{1,2,*}
\\
\bigskip
\textbf{1} Scientific Computing Group, Heidelberg Institute for Theoretical Studies, Heidelberg, Germany
\\
\textbf{2} Institute for Theoretical Informatics, Karlsruhe Institute of Technology, Karlsruhe, Germany
\\
\bigskip

% Insert additional author notes using the symbols described below. Insert symbol callouts after author names as necessary.
%
% Remove or comment out the author notes below if they aren't used.
%
% Primary Equal Contribution Note
% \Yinyang These authors contributed equally to this work.

% Additional Equal Contribution Note
% Also use this double-dagger symbol for special authorship notes, such as senior authorship.
% \ddag These authors also contributed equally to this work.

% Current address notes
% \textcurrency Current Address: Dept/Program/Center, Institution Name, City, State, Country
% change symbol to "\textcurrency a" if more than one current address note
% \textcurrency b Insert second current address
% \textcurrency c Insert third current address

% Deceased author note
% \dag Deceased

% Group/Consortium Author Note
% \textpilcrow Membership list can be found in the Acknowledgments section.

% Use the asterisk to denote corresponding authorship and provide email address in note below.
* \{lucas.czech,alexandros.stamatakis\}@h-its.org

\end{flushleft}

% ######################################################################################################################
%         Acronyms. Have to be inside the document environment...
% ######################################################################################################################

\begin{acronym}
    \ldots
\end{acronym}

% ######################################################################################################################
%         Abstract
% ######################################################################################################################

% >>> Remark for Editing:
% \vspace*{1em}
% At Alexis: Use commands \texttt{\textbackslash todo} for \todo{my comments},
% and \texttt{\textbackslash alexis} for \alexis{your comments}.
% <<< End of Remark.

% Please keep the abstract below 300 words
\section*{Abstract}

The exponential decrease in molecular sequencing cost generates unprecedented amounts of data.
Hence, scalable methods to analyze these data are required. %visualize and interpret such data.
Phylogenetic (or Evolutionary) Placement methods identify the evolutionary provenance of anonymous
sequences with respect to a given reference phylogeny.
This increasingly popular method is deployed for scrutinizing metagenomic samples
from environments such as water, soil, or the human gut.

Here, we present novel and, more importantly, highly scalable methods for analyzing phylogenetic placements of metagenomic samples.
% prepare sequence data for phylogenetic placement and to visualize and interpret the placement results.
More specifically, we introduce methods %visualizing and interpreting the placements:
% (a)
for visualizing differences between samples and their correlation with associated meta-data on the reference phylogeny,
% (b) measuring distances between samples based on their respective placements and
% (c)
as well as for clustering similar samples using a variant of the $k$-means method.
% to find and visualize correlation of the placements with the meta-data of the samples,
% to compare and cluster similar samples

To demonstrate the scalability and utility of our methods,
as well as to provide exemplary interpretations of our methods,
we applied them to \num{3} publicly available datasets comprising \num{9 782} samples
with a total of approximately \num{168} million sequences.
The results indicate that new biological insights can be attained via our methods.

% Author Summary
% The Author Summary is a 150-200 word non-technical summary of the work. Subject to editorial review, this short text
% is published with all research articles as a highlighted text box.
%
% Distinct from the scientific abstract, the Author Summary is included in the article to make findings accessible to
% an audience of both scientists and non-scientists. Ideally aimed to a level of understanding of an undergraduate student,
% the significance of the work should be presented simply, objectively, and without exaggeration.
%
% Authors should avoid the use of acronyms and complex scientific terms and write the text using a first-person voice.
% Authors may benefit from consulting with a science writer or press officer to ensure they effectively communicate their
% findings to a general audience.

% Please keep the Author Summary between 150 and 200 words
% Use first person. PLOS ONE authors please skip this step.
% Author Summary not valid for PLOS ONE submissions.
% \section*{Author summary}
% \todo{Do we need this?}

\linenumbers

% ######################################################################################################################
%         Introduction
% ######################################################################################################################

\section*{Introduction}
\label{sec:Introduction}
% I'm tempted to call this section ``Prologue'', to get one more name starting with a P...

\ldots


Here, we present novel, scalable methods to analyze and visualize phylogenetic placement data.
The remainder of this article is structured as follows.
First, we introduce the necessary terminology,
and provide an overview over existing post-analysis methods for phylogenetic placements.
Then, we describe several novel methods for visualizing differences between the placement data of distinct environmental samples,
and for visualizing their correlation with per-sample meta-data.
Furthermore, we propose a clustering algorithm for samples that is useful for handling extremely large environmental studies.
Lastly, we apply our methods to three real word datasets,
namely, \acf{BV} \cite{Srinivasan2012}, \acf{TO} \cite{Karsenti2011,Sunagawa2015,Guidi2016}, and \acf{HMP} \cite{Huttenhower2012,Methe2012},
which are introduced in more detail later.
We provide exemplary interpretations of the results obtaing from these datasets,
compare the results to existing methods, and analyze the run-time performance of our methods.

Our methods are implemented in our tool \toolname{gappa},
available under GPLv3 at \url{http://github.com/lczech/gappa}.
We provide an overview of the tool in S2 Text. % Supplementary Section \nameref{supp:sec:PipelineImplementation}.
Furthermore, scripts, data, and other tools used for the tests and generating the figures presented here
are available at \url{http://github.com/lczech/placement-methods-paper}.

% ======================================================================================================================
%     Phylogenetic Placement
% ======================================================================================================================

\subsection*{Phylogenetic placement}
\label{sec:Introduction:sub:PhylogeneticPlacement}

\ldots

% ======================================================================================================================
%     Edge Masses
% ======================================================================================================================

\subsection*{Edge masses}
\label{sec:Introduction:sub:Masses}

\ldots

% ======================================================================================================================
%     Edge Imbalances
% ======================================================================================================================

\subsection*{Edge imbalances}
\label{sec:Introduction:sub:Imbalances}

\ldots

% ######################################################################################################################
%         Methods
% ######################################################################################################################

\section*{Methods}
\label{sec:MaterialsMethods}

In this section, we introduce novel methods for analyzing and visualizing
the phylogenetic placement data of a set of environmental samples.

We initially describe the methods.
We then assess their application to real world data and their computational efficiency in Section \nameref{sec:Results}.

% ======================================================================================================================
%     Visualization
% ======================================================================================================================

\subsection*{Visualization}
\label{sec:MaterialsMethods:sub:Visualization}

% ----------------------------------------------------------------------------------------------------------------------
%     Edge Dispersion
% ----------------------------------------------------------------------------------------------------------------------

\subsubsection*{Edge Dispersion}
\label{sec:MaterialsMethods:sub:Visualization:sub:Dispersion}

\ldots

% ----------------------------------------------------------------------------------------------------------------------
%     Edge Correlation
% ----------------------------------------------------------------------------------------------------------------------

\subsubsection*{Edge Correlation}
\label{sec:MaterialsMethods:sub:Visualization:sub:Correlation}

\ldots

% ======================================================================================================================
%     Distance Measures
% ======================================================================================================================

% \subsection*{Distance Measures}
% \label{sec:MaterialsMethods:sub:DistanceMeasures}

% Removed Node Histgram Distance from the paper.

% ======================================================================================================================
%     Clustering
% ======================================================================================================================

\subsection*{Clustering}
\label{sec:MaterialsMethods:sub:Clustering}


% ----------------------------------------------------------------------------------------------------------------------
%     k-means for Phylogenetic Samples
% ----------------------------------------------------------------------------------------------------------------------

% PDF bookmarks do not accept the $k$ math mode, so make it use the letter instead.
\subsubsection*{Phylogenetic \texorpdfstring{$k$-means}{k-means}}
\label{sec:MaterialsMethods:sub:Clustering:sub:kmeans}


% ----------------------------------------------------------------------------------------------------------------------
%     Algorithmic Improvements
% ----------------------------------------------------------------------------------------------------------------------

\subsubsection*{Algorithmic improvements}
\label{sec:MaterialsMethods:sub:Clustering:sub:Improvements}

% ----------------------------------------------------------------------------------------------------------------------
%     Edge Imbalances
% ----------------------------------------------------------------------------------------------------------------------

\subsubsection*{Imbalance \texorpdfstring{$k$-means}{k-means}}
\label{sec:MaterialsMethods:sub:Clustering:sub:EdgeImbalances}

% ######################################################################################################################
%         Results
% ######################################################################################################################

% Results and Discussion can be combined.
\section*{Results and discussion}
\label{sec:Results}

% ======================================================================================================================
%     Visualization
% ======================================================================================================================

\subsection*{Visualization}
\label{sec:Results:sub:Visualization}

% ======================================================================================================================
%     Clustering
% ======================================================================================================================

\subsection*{Clustering}
\label{sec:Results:sub:Clustering}


% ######################################################################################################################
%         Discussion
% ######################################################################################################################

% \section*{Discussion}
% \label{sec:Discussion}
%
% Results and Discussion are be combined, as allowed by PLOS.

% ######################################################################################################################
%         Conclusion
% ######################################################################################################################

\section*{Conclusion}
\label{sec:Conclusion}

We presented novel, scalable methods to analyze and visualize phylogenetic placements.

The presented methods take either the edge masses or the edge imbalances as input,
and can hence analyze different aspects of the placements.
While edge masses reveal information about single branches,
edge imbalances take entire \acl{RT} clades into account.
Depending on the task at hand, either of them might be preferable,
although they generally exhibit similar properties.
We emphasize again the importance of appropriately normalizing the sample sizes as required.
That is, depending on the type of sequence data,
using either absolute or relative abundances is critical to allow for meaningful interpretation of the results.

We tested our novel methods on three real-world datasets and gave exemplary interpretations of the results.
We further showed that these results are consistent with existing methods as well as empirical biological knowledge.
Hence, our methods will also be useful to unravel new, unexplored relationships in metagenomic data.

The methods are computationally inexpensive, and are thus, as we have demonstrated, applicable to large datasets.
They are implemented in our tool \toolname{gappa},
which is freely available under GPLv3 at \url{http://github.com/lczech/gappa}.
% In Supplementary Section~\ref{supp:sec:Implementation}, we briefly introduce the software and its commands.
Furthermore, scripts, data and other tools used for the tests and figures presented here
are available at \url{http://github.com/lczech/placement-methods-paper}.

% ######################################################################################################################
%         Acknowledgments
% ######################################################################################################################

\section*{Acknowledgments}
\label{sec:Acknowledgments}

% ######################################################################################################################
%         References
% ######################################################################################################################

% Either type in your references using
% \begin{thebibliography}{}
% \bibitem{}
% Text
% \end{thebibliography}
%
% or
%
% Compile your BiBTeX database using our plos2015.bst
% style file and paste the contents of your .bbl file
% here. See http://journals.plos.org/plosone/s/latex for
% step-by-step instructions.
%
% \begin{thebibliography}{10}
%
% \bibitem{bib1}
% Conant GC, Wolfe KH.
% \newblock {{T}urning a hobby into a job: how duplicated genes find new
%   functions}.
% \newblock Nat Rev Genet. 2008 Dec;9(12):938--950.
%
% \bibitem{bib2}
% Ohno S.
% \newblock Evolution by gene duplication.
% \newblock London: George Alien \& Unwin Ltd. Berlin, Heidelberg and New York:
%   Springer-Verlag.; 1970.
%
% \bibitem{bib3}
% Magwire MM, Bayer F, Webster CL, Cao C, Jiggins FM.
% \newblock {{S}uccessive increases in the resistance of {D}rosophila to viral
%   infection through a transposon insertion followed by a {D}uplication}.
% \newblock PLoS Genet. 2011 Oct;7(10):e1002337.
%
% \end{thebibliography}

\nolinenumbers

% \clearpage
\bibliography{pppp-bib}

% ######################################################################################################################
%         Supplement
% ######################################################################################################################

% ######################################################################################################################
%         Supplement Texts
% ######################################################################################################################

\pagebreak
\beginsupplement

\section*{Supporting information}

\paragraph{S1 Text.}
\label{sec:supp:S1Text}

Empirical datasets

\paragraph{S2 Text.}
\label{sec:supp:S2Text}

Pipeline and implementation

% ######################################################################################################################
%         Supplement Tables
% ######################################################################################################################

% ######################################################################################################################
%         Supplement Figures
% ######################################################################################################################

% ######################################################################################################################
%     Original Template Examples
% ######################################################################################################################

% \paragraph*{S1 Fig.}
% \label{S1_Fig}
% {\bf Bold the title sentence.} Add descriptive text after the title of the item (optional).

% \paragraph*{S2 Fig.}
% \label{S2_Fig}
% {\bf Lorem ipsum.} Maecenas convallis mauris sit amet sem ultrices gravida. Etiam eget sapien nibh. Sed ac ipsum eget enim egestas ullamcorper nec euismod ligula. Curabitur fringilla pulvinar lectus consectetur pellentesque.

% \paragraph*{S1 File.}
% \label{S1_File}
% {\bf Lorem ipsum.}  Maecenas convallis mauris sit amet sem ultrices gravida. Etiam eget sapien nibh. Sed ac ipsum eget enim egestas ullamcorper nec euismod ligula. Curabitur fringilla pulvinar lectus consectetur pellentesque.

% \paragraph*{S1 Video.}
% \label{S1_Video}
% {\bf Lorem ipsum.}  Maecenas convallis mauris sit amet sem ultrices gravida. Etiam eget sapien nibh. Sed ac ipsum eget enim egestas ullamcorper nec euismod ligula. Curabitur fringilla pulvinar lectus consectetur pellentesque.

% \paragraph*{S1 Appendix.}
% \label{S1_Appendix}
% {\bf Lorem ipsum.} Maecenas convallis mauris sit amet sem ultrices gravida. Etiam eget sapien nibh. Sed ac ipsum eget enim egestas ullamcorper nec euismod ligula. Curabitur fringilla pulvinar lectus consectetur pellentesque.

% \paragraph*{S1 Table.}
% \label{S1_Table}
% {\bf Lorem ipsum.} Maecenas convallis mauris sit amet sem ultrices gravida. Etiam eget sapien nibh. Sed ac ipsum eget enim egestas ullamcorper nec euismod ligula. Curabitur fringilla pulvinar lectus consectetur pellentesque.

% \pagebreak
% \todo{Remove this page break!}

\end{document}

